### DISCUSSION TITLE
De-duplication / Removing recipients from a string of CC'd addresses

### DISCUSSION THREAD

**Discussion by endoflevelboss** [rspamd/rspamd#5182] [General]
I'm trying to implement a set of de-duplication rules for incoming mails. We are running a ticket system that hides the internal structure of our organisation from the outside world. We do not publish personal contact emails and all task assignment is done internally in the helpdesk system. The actual communication technically goes through a unified single mail address. However, for historical reasons, personal mail addresses are still circulating and a significant share of incoming mails is sent to nearly every mail address ever disclosed. 

My current strategy is removing extra recipients based on a set of sieve rules. Everything that includes our mail gate address can't be a private communication by definition, so I'm dropping every email that would go to any personal inbox while having a group inbox set on TO or CC. That really works great and I successfully shut the door on duplicates being delivered to personal inboxes. However, the mail that finally goes into our ticket system still contains the full list of recipients as set by the sender. If one forgets to clean this up on the first response (or if the sender adds new recipients later), this can create crazy loops.

Long story short, I would like to fix the incoming mail upon first reception by cleaning the TO and CC headers before the message actually is delivered to the ticket system's inbox.

Is there a way to implement such a scheme in rspamd to facilitate such a de-duplication rule set? The scheme would look like: Loop over everybody on TO and CC and check if a well known group email is included. If so, clean TO and CC of all further recipients in my own mail domain only. Keep all external recipients, no matter if they are on TO or CC. 

Every hint or advice is greatly appreciated! Thank you in advance!

**Comment by fatalbanana** [comment]
You can call [task:set_milter_reply](https://rspamd.com/doc/lua/rspamd_task.html#m70081) to modify headers supposing your MTA is able to follow such instructions (e.g. Postfix); meaning of the numeric arguments in `remove_headers` is explained [in milter_headers module documentation](https://rspamd.com/doc/modules/milter_headers.html#removing-headers).

Modifying these headers is likely to break digital signatures (like DKIM).

~~~lua
-- /etc/rspamd/rspamd.local.lua
-- quick and dirty example; display names are lost
local rspamd_util = require 'rspamd_util'

rspamd_config.MANGLE_RECIPIENT_INFORMATION = {
  callback = function(task)
    -- decide if we should apply mangling
    if not rspamd_util.strequal_caseless(task:get_principal_recipient(), 'support@example.net') then return end
    -- get recipients from To/Cc ...
    -- to handle these individually use rspamd_util.parse_mail_address with task:get_header
    local mime_rcpts = task:get_recipients('mime')
    local keep_rcpts = {}
    for _, rcpt in ipairs(mime_rcpts) do
      if rcpt.domain ~= 'example.net' then
        table.insert(keep_rcpts, rcpt.addr)
      end
    end
    local new_to_hdr = table.concat(keep_rcpts, ', ')
   -- rewrite To header, strip Cc
    task:set_milter_reply{
      add_headers = {
        To = new_to_hdr,
      },
      remove_headers = {
        To = 0,
        Cc = 0,
      }
    }
  end,
  type = 'idempotent'
}
~~~

**Comment by endoflevelboss** [comment]
Thank you so much for this helpful and lighning fast response! I'll try my luck based on the pointers you gave me.

Just to be sure, am I assuming correctly that expectedly breaking DKIM would only be an issue if the modification is done after this very milter is applied and that there is a way to define the order in which milters are called? Am I missing another possible side-effect?

**Reply by fatalbanana** [reply]
DKIM will work but the email as delivered would not pass DKIM.

**Reply by endoflevelboss** [reply]
I understand that implication. Still wondering if that would mean a problem. 

Am I assuming correctly that the original, unaltered message would successfully pass the DKIM milter, then would next be modified by my de-duplication milter (DKIM gets broken) and finally is delivered to the inbox from where it will be further processed by the ticket system. As long as the ticket system relies on the added X-header (DKIM confirmed ok) I don‘t see why the invalidated DKIM would be an issue. Am I missing something?
